contracts/contracts/sbtc-withdrawal.clar

```diff
@@ -33,6 +45,110 @@
   )
 )
 
+;; Accept a withdrawal request
+(define-public (accept-withdrawal-request (request-id uint) 
+                                          (bitcoin-txid (buff 32)) 
+                                          (signer-bitmap uint)
+                                          (output-index uint)
+                                          (fee uint))
+  (let 
+    (
+      (current-signer-data (contract-call? .sbtc-registry get-current-signer-data))   
+      (request  (unwrap! (contract-call? .sbtc-registry get-withdrawal-request request-id) ERR_INVALID_REQUEST))
+      (requested-max-fee (get max-fee request))
+      (requester (get sender request))
+    )
+      ;; Check that the caller is the current signer principal
+      (asserts! (is-eq (get current-signer-principal current-signer-data) tx-sender) ERR_INVALID_CALLER)
+
+      ;; Check whether it was already accepted or rejected
+      (asserts! (is-none (get status request)) ERR_ALREADY_PROCESSED)
+
+      ;; Check that fee is not higher than requesters max fee
+      (asserts! (<= fee requested-max-fee) ERR_FEE_TOO_HIGH)
+
+      ;; Burn the locked-sbtc
+      (try! (contract-call? .sbtc-token protocol-burn-locked (get amount request) requester))
+
+      ;; Mint the difference b/w max-fee of the request & fee actually paid back to the user in sBTC
+      (if (is-eq (- requested-max-fee fee) u0)
+        true
+        (try! (contract-call? .sbtc-token protocol-mint (- requested-max-fee fee) requester))
+      )
+
+      ;; Call into registry to confirm accepted withdrawal
+      (try! (contract-call? .sbtc-registry complete-withdrawal request-id true (some bitcoin-txid) (some signer-bitmap) (some output-index) (some fee)))
+
+      (ok true)
+  )
+)
+
+;; Reject a withdrawal request
+(define-public (reject-withdrawal (request-id uint) (signer-bitmap uint))
+  (let
+     (
+      (current-signer-data (contract-call? .sbtc-registry get-current-signer-data))   
+      (withdrawal (unwrap! (contract-call? .sbtc-registry get-withdrawal-request request-id) ERR_INVALID_REQUEST))
+     )
+
+    ;; Check that the caller is the current signer principal
+    (asserts! (is-eq (get current-signer-principal current-signer-data) tx-sender) ERR_INVALID_CALLER)
+
+    ;; Check that request status is currently-pending
+    (asserts! (is-none (get status withdrawal)) ERR_ALREADY_PROCESSED)
+
+    ;; Burn sbtc-locked & re-mint sbtc to original requester
+    (try! (contract-call? .sbtc-token protocol-unlock (get amount withdrawal) (get sender withdrawal)))
+
+    ;; Call into registry to confirm accepted withdrawal
+    (try! (contract-call? .sbtc-registry complete-withdrawal request-id false none none none none))
+
+    (ok true)
+  )
+)
+;; Reject multiple withdrawal requests
+(define-public (complete-withdrawals (withdrawals (list 100 {request-id: uint, status: bool, signer-bitmap: uint, bitcoin-txid: (optional (buff 32)), output-index: (optional uint), fee: (optional uint)})))
```

Comment:
I think we can simplify this functionality (or break it up into `complete-accept-withdrawals` and `complete-reject-withdrawals` or whatever you want to call it). The signers need to wait for a bitcoin confirmation to complete an accepted withdrawal, but they can complete a rejection right away (well, right after "voting"). So bundling acceptances makes sense, we find out about them in bundles when the transaction package gets confirmed on bitcoin. The flow for rejections should be much faster, since it would happen shortly after it gets confirmed on a stacks Nakamoto  block.

Also, I think I prefer breaking them out since it'll be easier to reason about both here and in the signer codebase.

Reply:
Gotcha. Hmmm, my thinking here is giving the signers flexibility to bundle both accepts & rejects together to submit a single list & therefore transaction calls. Which, while separating them might improve readability, the real pro-argument here that is signers have to make more transactions, which means they have to pay more fees.

Reply:
Ahh right, good point.

There is a decent chance that having it as one function will not lead to fewer transaction calls. We might have signers make a contract call for rejects every ~5-30 seconds[^1] and then make another for the accepts every ~10 minutes (for new bitcoin blocks) since this approach has the better user experience.

---

Anyway, I'm okay leaving this as is for now. We can always revisit this later when we get this part nailed down in the signer.

[^1]: The signers might reject shortly after they find out about it, which is when the request is included in a new Nakamoto block. I'm not sure how quickly Nakamoto blocks are supposed to be created.